pkg/prober/prober.go

```diff
@@ -0,0 +1,160 @@
+package prober
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/component"
+	"github.com/prometheus/common/route"
+)
+
+const (
+	healthyEndpointPath  = "/-/healthy"
+	readyEndpointPath    = "/-/ready"
+	okProbeFmt           = "thanos %v is %v"
+	errorProbeFmt        = "thanos %v is not %v. Reason: %v"
+	probeErrorHTTPStatus = 503
+	initialErrorFmt      = "thanos %s is initializing"
+)
+
+// Prober represents health and readiness status of given component.
+type Prober struct {
+	logger             log.Logger
+	componentMtx       sync.RWMutex
+	component          component.Component
+	readyMtx           sync.RWMutex
+	readiness          error
+	healthyMtx         sync.RWMutex
+	healthiness        error
+	readyStateMetric   prometheus.Gauge
+	healthyStateMetric prometheus.Gauge
+}
+
+// NewProber returns Prober representing readiness and healthiness of given component.
+func NewProber(component component.Component, logger log.Logger, reg prometheus.Registerer) *Prober {
+	initialErr := fmt.Errorf(initialErrorFmt, component)
+
+	p := &Prober{
+		component:   component,
+		logger:      logger,
+		healthiness: initialErr,
+		readiness:   initialErr,
+		readyStateMetric: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: fmt.Sprintf("thanos_%s_ready", component),
```

Comment:
TBH instead of that I would just do `prober_ready",  and then:
* have `label` component. This will allow for better, potential aggregations in future.
* you can have just `prober_ready` and do not specify `thanos` prefix. This allows to reuse this package outside of Thanos even (why not)!

Thanos can then on caller side do `prometheus.WrapRegistererWithPrefix("thanos"` (: 

What do you think?

same for `prober_healthy`

Reply:
I was thinking about the naming a lot bud than looked to the rest of packages and found everywhere hardcoded the thanos prefix so I wanted to stay consistent but I agree your suggestion is more variable.

Applied as suggested

Reply:
Yes, we did not think about that in initial implementation, but now I would aim for no `thanos_` prefix (: